drivers/i3c/i3c_nxp_s32.c

```diff
@@ -0,0 +1,491 @@
+
+#define NXP_S32_I3C_DEFINE_CALLBACK(n)							\
+	void nxp_s32_i3c_##n##_master_callback(I3c_Ip_MasterEventType Event)		\
+	{										\
+		const struct device *dev = DEVICE_DT_GET(NXP_S32_I3C_NODE(n));		\
+											\
+		nxp_s32_i3c_master_callback(dev, Event);				\
+	}
+
+#define NXP_S32_I3C_S32_CONFIG(n)							\
+	static I3c_Ip_MasterStateType nxp_s32_i3c_##n##_state =				\
+	{										\
+		0U,									\
+		NULL_PTR,								\
+		NULL_PTR,								\
+		I3C_IP_STATUS_SUCCESS,							\
+		{									\
+			0x00,								\
+			(boolean)FALSE,							\
+			I3C_IP_WRITE,							\
+			I3C_IP_TRANSFER_BYTES,						\
+			I3C_IP_BUS_TYPE_I2C						\
+		},									\
+		I3C_IP_USING_INTERRUPTS,						\
+		(boolean)FALSE,								\
+		(I3c_Ip_MasterCallbackType)nxp_s32_i3c_##n##_master_callback,		\
+	};										\
+											\
+	static const I3c_Ip_MasterConfigType nxp_s32_i3c_##n##_config =			\
+	{										\
+		I3C_IP_MASTER_ON,							\
+		(boolean)TRUE,								\
+		I3C_IP_MASTER_HIGH_KEEPER_NONE,						\
+		0U,									\
+		0U,									\
+		0U,									\
+		0U,									\
+		(boolean)FALSE,								\
+		0U,									\
+		(boolean)FALSE,								\
+		(boolean)FALSE,								\
+		&nxp_s32_i3c_##n##_state,						\
+	};
+
+#define NXP_S32_I3C_INIT_DEVICE(n)							\
+	NXP_S32_I3C_DEFINE_CALLBACK(n)							\
+	NXP_S32_I3C_DECLARE_INTERRUPT(n)						\
+	NXP_S32_I3C_S32_CONFIG(n)							\
+	PINCTRL_DT_DEFINE(NXP_S32_I3C_NODE(n));						\
+	static struct nxp_s32_i3c_data nxp_s32_i3c_data_##n;				\
+	static struct i3c_device_desc s32_i3c_device_array[] =				\
+		I3C_DEVICE_ARRAY_DT(NXP_S32_I3C_NODE(n));				\
+	static struct i3c_i2c_device_desc s32_i3c_i2c_device_array[] =			\
+		I3C_I2C_DEVICE_ARRAY_DT(NXP_S32_I3C_NODE(n));				\
+	static const struct nxp_s32_i3c_config nxp_s32_i3c_config_##n = {		\
+		.instance	 = n,							\
+		.functional_clk  = DT_PROP(NXP_S32_I3C_NODE(n), clock_frequency),	\
+		.i3c_master_cfg  = (I3c_Ip_MasterConfigType *)&nxp_s32_i3c_##n##_config,\
+		.device_list.i3c = s32_i3c_device_array,				\
+		.device_list.num_i3c = ARRAY_SIZE(s32_i3c_device_array),		\
+		.device_list.i2c = s32_i3c_i2c_device_array,				\
+		.device_list.num_i2c = ARRAY_SIZE(s32_i3c_i2c_device_array),		\
+		.pincfg = PINCTRL_DT_DEV_CONFIG_GET(NXP_S32_I3C_NODE(n)),		\
+		NXP_S32_I3C_CONFIG_INTERRUPT(n)						\
+	};										\
+	static struct nxp_s32_i3c_data nxp_s32_i3c_data_##n = {				\
+		.ctrl_config = {							\
+			.scl = {							\
+				.i3c = DT_PROP_OR(NXP_S32_I3C_NODE(n), i3c_scl_hz, 0),	\
+				.i2c = DT_PROP_OR(NXP_S32_I3C_NODE(n), i2c_scl_hz, 0),	\
+			},								\
+		},									\
+		.i3c_od_scl_hz = DT_PROP_OR(NXP_S32_I3C_NODE(n), i3c_od_scl_hz, 0),	\
+	};										\
+	DEVICE_DT_DEFINE(NXP_S32_I3C_NODE(n),						\
+			nxp_s32_i3c_init,						\
+			NULL,								\
+			&nxp_s32_i3c_data_##n,						\
+			&nxp_s32_i3c_config_##n,					\
+			POST_KERNEL,							\
+			CONFIG_I3C_CONTROLLER_INIT_PRIORITY,				\
+			&nxp_s32_i3c_driver_api);
+
+#if DT_NODE_HAS_STATUS(NXP_S32_I3C_NODE(0), okay)
```

Comment:
use `DT_INST_FOREACH_STATUS_OKAY`

Reply:
This macro is used because the module instance is needed for the HAL, already discussed in https://github.com/zephyrproject-rtos/zephyr/pull/49161#discussion_r949770066